subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/execution/ProgramParser.kt

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.execution
+
+import org.gradle.kotlin.dsl.support.compilerMessageFor
+
+
+internal
+object ProgramParser {
+
+    fun parse(source: ProgramSource, kind: ProgramKind, target: ProgramTarget): Program = try {
+        programFor(source, kind, target)
+    } catch (unexpectedBlock: UnexpectedBlock) {
+        handleUnexpectedBlock(unexpectedBlock, source.text, source.path)
+    }
+
+    private
+    fun programFor(source: ProgramSource, kind: ProgramKind, target: ProgramTarget): Program {
+
+        val topLevelBlockIds =
+            when (target) {
+                ProgramTarget.Project -> arrayOf("buildscript", "plugins")
+                ProgramTarget.Settings -> arrayOf("buildscript", "pluginManagement")
+                ProgramTarget.Gradle -> arrayOf("initscript")
+            }
+
+        val (comments, topLevelBlocks) = lex(source.text, *topLevelBlockIds)
+
+        checkForSingleBlocksOf(topLevelBlockIds, topLevelBlocks)
+
+        val sourceWithoutComments =
+            source.map { it.erase(comments) }
+
+        val buildscriptFragment =
+            topLevelBlocks
+                .singleSectionOf(topLevelBlockIds[0])
+                ?.let { sourceWithoutComments.fragment(it) }
+
+        val pluginsFragment =
+            topLevelBlocks
+                .takeIf { target == ProgramTarget.Project && kind == ProgramKind.TopLevel }
+                ?.singleSectionOf("plugins")
+                ?.let { sourceWithoutComments.fragment(it) }
+
+        val buildscript =
+            buildscriptFragment?.takeIf { it.isNotBlank() }?.let(Program::Buildscript)
+
+        val plugins =
+            pluginsFragment?.takeIf { it.isNotBlank() }?.let(Program::Plugins)
+
+        val stage1 =
+            buildscript?.let { bs ->
+                plugins?.let { ps ->
+                    Program.Stage1Sequence(bs, ps)
+                } ?: bs
+            } ?: plugins
```

Comment:
Took me a few seconds to figure out what was going on here. But the end result is quite clever.
It takes a bit of time to start reading `?.let {` as an if statement.